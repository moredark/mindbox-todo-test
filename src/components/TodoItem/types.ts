import { Todo } from '../../types';

export interface TodoItemProps {
  todo: Todo;
  onToggle: (id: string) => void;
  onDelete: (id: string) => void;
}
