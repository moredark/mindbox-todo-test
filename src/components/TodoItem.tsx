import React from 'react';
import { CheckBox, Box, Button, Text } from 'grommet';
import { Trash } from 'grommet-icons';
import styled from 'styled-components';
import { Todo } from '../types';

interface TodoItemProps {
  todo: Todo;
  onToggle: (id: string) => void;
  onDelete: (id: string) => void;
}

const TodoText = styled(Text)<{ completed: boolean }>`
  text-decoration: ${props => props.completed ? 'line-through' : 'none'};
  opacity: ${props => props.completed ? 0.7 : 1};
  flex: 1;
`;

export const TodoItem: React.FC<TodoItemProps> = ({ todo, onToggle, onDelete }) => {
  return (
    <Box
      direction="row"
      align="center"
      pad="small"
      margin={{ vertical: 'xsmall' }}
      background="background-contrast"
      round="small"
      border={{ color: 'dark-3', size: '1px' }}
    >
      <CheckBox
        checked={todo.completed}
        onChange={() => onToggle(todo.id)}
        toggle={false}
        data-testid={`todo-checkbox-${todo.id}`}
      />
      <TodoText 
        completed={todo.completed}
        data-testid={`todo-text-${todo.id}`}
        color={todo.completed ? "dark-5" : "text"}
        margin={{ left: 'small' }}
      >
        {todo.text}
      </TodoText>
      <Button
        icon={<Trash size="small" color="status-critical" />}
        onClick={() => onDelete(todo.id)}
        plain
        data-testid={`todo-delete-${todo.id}`}
      />
    </Box>
  );
};
