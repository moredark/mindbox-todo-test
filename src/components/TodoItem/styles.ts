import styled from 'styled-components';
import { Box, Text } from 'grommet';
import { slideIn, slideOut } from '../../styles/animations';
import { ANIMATION_DURATION } from '../../constants/animations';

export const AnimatedTodoBox = styled(Box)`
  animation: ${slideIn} ${ANIMATION_DURATION.SHORT}ms ease-out;
  &.removing {
    animation: ${slideOut} ${ANIMATION_DURATION.SHORT}ms ease-out forwards;
  }
`;

export const TodoText = styled(Text)<{ completed: boolean }>`
  text-decoration: ${props => props.completed ? 'line-through' : 'none'};
  opacity: ${props => props.completed ? 0.7 : 1};
  flex: 1;
  transition: all ${ANIMATION_DURATION.SHORT}ms ease;
`;
