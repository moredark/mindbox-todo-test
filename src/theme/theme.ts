import { ThemeType } from 'grommet';
import { deepMerge } from 'grommet/utils';
import { dark } from 'grommet/themes';

export const accentColor = '#7D4CDB';

export const customTheme: ThemeType = deepMerge(dark, {
  global: {
    colors: {
      background: '#121212',
      'background-back': '#1E1E1E',
      'background-front': '#2D2D2D',
      'background-contrast': '#3D3D3D',
      text: '#EEEEEE',
      'text-strong': '#FFFFFF',
      'dark-6': '#999999',
      'dark-5': '#BBBBBB',
      'accent-1': accentColor,
      brand: accentColor
    },
    focus: {
      border: {
        color: accentColor
      }
    }
  },
  button: {
    border: {
      radius: '4px'
    },
    primary: {
      color: accentColor
    }
  },
  checkBox: {
    hover: {
      border: {
        color: accentColor
      }
    },
    color: accentColor
  }
});
