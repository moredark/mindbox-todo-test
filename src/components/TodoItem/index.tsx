import React from 'react';
import { CheckBox, Button } from 'grommet';
import { Trash } from 'grommet-icons';
import { TodoItemProps } from './types';
import { AnimatedTodoBox, TodoText } from './styles';
import { ANIMATION_DURATION } from '../../constants/animations';

export const TodoItem: React.FC<TodoItemProps> = ({ todo, onToggle, onDelete }) => {
  const [isRemoving, setIsRemoving] = React.useState(false);

  const handleDelete = () => {
    setIsRemoving(true);
    setTimeout(() => {
      onDelete(todo.id);
    }, ANIMATION_DURATION.SHORT);
  };

  return (
    <AnimatedTodoBox
      direction="row"
      align="center"
      pad="small"
      margin={{ vertical: 'xsmall' }}
      background="background-contrast"
      round="small"
      border={{ color: 'dark-3', size: '1px' }}
      className={isRemoving ? 'removing' : ''}
    >
      <CheckBox
        checked={todo.completed}
        onChange={() => onToggle(todo.id)}
        data-testid={`todo-checkbox-${todo.id}`}
      />
      <TodoText 
        completed={todo.completed}
        data-testid={`todo-text-${todo.id}`}
        color={todo.completed ? "dark-5" : "text"}
        margin={{ left: 'small' }}
      >
        {todo.text}
      </TodoText>
      <Button
        icon={<Trash size="small" color="status-critical" />}
        onClick={handleDelete}
        plain
        data-testid={`todo-delete-${todo.id}`}
      />
    </AnimatedTodoBox>
  );
};
