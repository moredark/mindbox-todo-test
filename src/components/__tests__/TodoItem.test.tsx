import { render, screen, fireEvent } from '@testing-library/react';
import { Grommet } from 'grommet';
import { TodoItem } from '../TodoItem';
import { customTheme } from '../../theme/theme';

describe('TodoItem', () => {
  const mockTodo = {
    id: '1',
    text: 'Test todo',
    completed: false,
    createdAt: new Date('2024-01-01')
  };

  const defaultProps = {
    todo: mockTodo,
    onToggle: jest.fn(),
    onDelete: jest.fn()
  };

  const renderComponent = (props = defaultProps) => {
    return render(
      <Grommet theme={customTheme}>
        <TodoItem {...props} />
      </Grommet>
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot for incomplete todo', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for completed todo', () => {
    const { container } = renderComponent({
      ...defaultProps,
      todo: { ...mockTodo, completed: true }
    });
    expect(container).toMatchSnapshot();
  });

  it('should call onToggle when checkbox is clicked', () => {
    renderComponent();
    const checkbox = screen.getByTestId(`todo-checkbox-${mockTodo.id}`);
    
    fireEvent.click(checkbox);
    
    expect(defaultProps.onToggle).toHaveBeenCalledTimes(1);
    expect(defaultProps.onToggle).toHaveBeenCalledWith(mockTodo.id);
  });

  it('should call onDelete when delete button is clicked', () => {
    renderComponent();
    const deleteButton = screen.getByTestId(`todo-delete-${mockTodo.id}`);
    
    fireEvent.click(deleteButton);
    
    expect(defaultProps.onDelete).toHaveBeenCalledTimes(1);
    expect(defaultProps.onDelete).toHaveBeenCalledWith(mockTodo.id);
  });

  it('should display todo text', () => {
    renderComponent();
    const todoText = screen.getByTestId(`todo-text-${mockTodo.id}`);
    expect(todoText).toHaveTextContent(mockTodo.text);
  });

  it('should apply completed styles when todo is completed', () => {
    renderComponent({
      ...defaultProps,
      todo: { ...mockTodo, completed: true }
    });
    
    const todoText = screen.getByTestId(`todo-text-${mockTodo.id}`);
    expect(todoText).toHaveStyle({
      textDecoration: 'line-through',
      opacity: '0.7'
    });
  });

  it('should apply normal styles when todo is not completed', () => {
    renderComponent();
    
    const todoText = screen.getByTestId(`todo-text-${mockTodo.id}`);
    expect(todoText).toHaveStyle({
      textDecoration: 'none',
      opacity: '1'
    });
  });
});
