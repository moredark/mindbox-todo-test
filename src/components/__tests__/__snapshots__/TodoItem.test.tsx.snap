// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should match snapshot for completed todo 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 lioFhE"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dKnYnC sc-VHjGu jdxga-D"
    >
      <label
        class="StyledCheckBox__StyledCheckBoxContainer-sc-1dbk5ju-1 ibppuy"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 kkuNuI StyledCheckBox-sc-1dbk5ju-6 eaiPwo"
        >
          <input
            checked=""
            class="StyledCheckBox__StyledCheckBoxInput-sc-1dbk5ju-2 kwqnhR"
            data-testid="todo-checkbox-1"
            type="checkbox"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 iRwvIO StyledCheckBox__StyledCheckBoxBox-sc-1dbk5ju-3 fQtXhU"
          >
            <svg
              class="StyledCheckBox__StyledCheckBoxIcon-sc-1dbk5ju-0 huNCKL"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 24 24"
            >
              <path
                d="M6,11.3 L10.3,16 L18,6.2"
                fill="none"
              />
            </svg>
          </div>
        </div>
      </label>
      <span
        class="StyledText-sc-1sadyjn-0 eZKRqX sc-gnOvAp eELmxq"
        data-testid="todo-text-1"
      >
        Test todo
      </span>
      <button
        class="StyledButton-sc-323bzc-0 kdwnZD"
        data-testid="todo-delete-1"
        type="button"
      >
        <svg
          aria-label="Trash"
          class="StyledIcon-sc-ofa7kd-0 jzlLJr"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 5h16v18H4V5zM1 5h22M9 1h6v4H9V1zm0 0h6v4H9V1zm6 8v10M9 9v10"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`TodoItem should match snapshot for incomplete todo 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 lioFhE"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 dKnYnC sc-VHjGu jdxga-D"
    >
      <label
        class="StyledCheckBox__StyledCheckBoxContainer-sc-1dbk5ju-1 ibppuy"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 kkuNuI StyledCheckBox-sc-1dbk5ju-6 eaiPwo"
        >
          <input
            class="StyledCheckBox__StyledCheckBoxInput-sc-1dbk5ju-2 kwqnhR"
            data-testid="todo-checkbox-1"
            type="checkbox"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 dmVmJK StyledCheckBox__StyledCheckBoxBox-sc-1dbk5ju-3 fQtXhU"
          />
        </div>
      </label>
      <span
        class="StyledText-sc-1sadyjn-0 iugIKr sc-gnOvAp gvMwTs"
        data-testid="todo-text-1"
      >
        Test todo
      </span>
      <button
        class="StyledButton-sc-323bzc-0 kdwnZD"
        data-testid="todo-delete-1"
        type="button"
      >
        <svg
          aria-label="Trash"
          class="StyledIcon-sc-ofa7kd-0 jzlLJr"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 5h16v18H4V5zM1 5h22M9 1h6v4H9V1zm0 0h6v4H9V1zm6 8v10M9 9v10"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
